### DISCUSSION TITLE
Cannot compile for Seeed-XIAO-BLE-Sense nRF52840

### DISCUSSION THREAD

**Discussion by BlueObsecurit** [pschatzmann/arduino-audio-tools#141] [Q&A] [answered]
Hi
I'm self taught, have done some things with arduino, but ultimately I've probably just forgot something stupid. Also first time I've used github so if I commit some faux pas, forgive me. 

Trying to compote the player sd analog example for https://www.seeedstudio.com/Seeed-XIAO-BLE-Sense-nRF52840-p-5253.html based on nrf52840 

Only thing I modify is adding a #define PIN_CS 7. I always get that AnalogAudioStream does not name a type. I've tried reinstating all relevant things, and make sure the compiler includes the .h file. Doesn't work and I'm stumped. Any suggestions?

**Comment by pschatzmann** [comment]
I am afraid that this processor is not supported.

**Reply by BlueObsecurit** [reply]
Yes, I figured out why nothing worked yesterday! Forgive me though, but it seems from AudioConfig.h that Nano 33 BLE is supported? Is this not the same processor?

I managed to get a pwm example compiled and it seems to stream, but the audio quality is so poor I can't quite tell if it is audio... this is 100% my fault though. I plan to get an I2S dac soon.

Unless you mean AnalogAudioStream is not supported for this processor which I understand now.

**Comment by pschatzmann** [comment, marked as answer]
Just a small pre-warning: I have added an I2S implementation for the Nano BLE Sense, but never had the chance to test it.
See https://github.com/pschatzmann/arduino-audio-tools/wiki/Supported-Architectures

**Reply by BlueObsecurit** [reply]
Well that's expected considering the scope and size of this library. Are you happy to work with me to attempt to get this working?
Currently I seem to have managed to include my board in the Nano 33 BLE i2s .h file, but I get this error:

Edit 2: my initial edit seems to have dropped. I am using player-sd-i2s with no modifications other than to get my processor added to the .h files.

```
src/AudioI2S/I2SStream.h:24:28: error: expected class-name before '{' token
 class I2S : public I2SBase {
```

And a bunch more that I am hoping are just related to that. I have tried looking around, unfortunatly I don't have enough experience with writing Arduino libraries to understand what the error means or what the suggested solutions mean.

Rest of the errors:

```
In file included (removed) from \src/AudioTools.h:17:0,
                 (removed) from \examples\examples-player\player-sd-i2s\player-sd-i2s.ino:13:
\src/AudioI2S/I2SStream.h:24:28: error: expected class-name before '{' token
 class I2S : public I2SBase {
                            ^
\src/AudioI2S/I2SStream.h: In destructor 'audio_tools::I2S<T>::~I2S()':
\src/AudioI2S/I2SStream.h:34:7: error: there are no arguments to 'end' that depend on a template parameter, so a declaration of 'end' must be available [-fpermissive]
       end();
       ^~~
\src/AudioI2S/I2SStream.h:34:7: note: (if you use '-fpermissive', G++ will accept your code, but allowing the use of an undeclared name is deprecated)
\src/AudioI2S/I2SStream.h: In member function 'bool audio_tools::I2S<T>::begin(audio_tools::I2SConfig)':
\src/AudioI2S/I2SStream.h:41:14: error: 'I2SBase' has not been declared
       return I2SBase::begin(cfg);
              ^~~~~~~
\src/AudioI2S/I2SStream.h: At global scope:
\src/AudioI2S/I2SStream.h:161:9: error: 'I2SBase' does not name a type; did you mean 'I2CName'?
         I2SBase i2s;
         ^~~~~~~
         I2CName
\src/AudioI2S/I2SStream.h: In member function 'audio_tools::I2SConfig audio_tools::I2SStream::defaultConfig(audio_tools::RxTxMode)':
\src/AudioI2S/I2SStream.h:83:20: error: 'i2s' was not declared in this scope
             return i2s.defaultConfig(mode);
                    ^~~
\src/AudioI2S/I2SStream.h: In member function 'virtual bool audio_tools::I2SStream::begin()':
\src/AudioI2S/I2SStream.h:87:20: error: 'i2s' was not declared in this scope
             return i2s.begin();
                    ^~~
\src/AudioI2S/I2SStream.h: In member function 'void audio_tools::I2SStream::begin(audio_tools::I2SConfig)':
\src/AudioI2S/I2SStream.h:93:13: error: 'i2s' was not declared in this scope
             i2s.begin(cfg);
             ^~~
\src/AudioI2S/I2SStream.h: In member function 'virtual void audio_tools::I2SStream::end()':
\src/AudioI2S/I2SStream.h:102:13: error: 'i2s' was not declared in this scope
             i2s.end();
             ^~~
\src/AudioI2S/I2SStream.h: In member function 'virtual void audio_tools::I2SStream::setAudioInfo(audio_tools::AudioBaseInfo)':
\src/AudioI2S/I2SStream.h:108:29: error: 'i2s' was not declared in this scope
             I2SConfig cfg = i2s.config();
                             ^~~
\src/AudioI2S/I2SStream.h: In member function 'virtual size_t audio_tools::I2SStream::write(const uint8_t*, size_t)':
\src/AudioI2S/I2SStream.h:125:20: error: 'i2s' was not declared in this scope
             return i2s.writeBytes(buffer, size);
                    ^~~
\src/AudioI2S/I2SStream.h: In member function 'virtual size_t audio_tools::I2SStream::readBytes(uint8_t*, size_t)':
\src/AudioI2S/I2SStream.h:130:20: error: 'i2s' was not declared in this scope
             return i2s.readBytes(data, length);
                    ^~~
\src/AudioI2S/I2SStream.h: In member function 'virtual int audio_tools::I2SStream::available()':
\src/AudioI2S/I2SStream.h:150:20: error: 'i2s' was not declared in this scope
             return i2s.available();
                    ^~~
\src/AudioI2S/I2SStream.h: In member function 'virtual int audio_tools::I2SStream::availableForWrite()':
\src/AudioI2S/I2SStream.h:155:20: error: 'i2s' was not declared in this scope
             return i2s.availableForWrite();
                    ^~~

```

**Comment by pschatzmann** [comment]
Which Board do you select: I am using -> Arduino Mbed OS Nano Boards ->  Arduino Nano 33 BLE.

When I compile I am getting lines like the following:
/Users/pschatzmann/Library/Arduino15/packages/arduino/tools/arm-none-eabi-gcc/7-2017q4/bin/arm-none-eabi-g++ -c -g3 -nostdlib @/Users/pschatzmann/Library/Arduino15/packages/arduino/hardware/mbed_nano/2.7.2/variants/ARDUINO_NANO33BLE/defines.txt @/Users/pschatzmann/Library/Arduino15/packages/arduino/hardware/mbed_nano/2.7.2/variants/ARDUINO_NANO33BLE/cxxflags.txt -DARDUINO_ARCH_NRF52840 -MMD -mcpu=cortex-m4 -mfloat-abi=softfp -mfpu=fpv4-sp-d16 -DARDUINO=10813 -DARDUINO_ARDUINO_NANO33BLE -DARDUINO_ARCH_MBED_NANO -DARDUINO_ARCH_MBED -DARDUINO_LIBRARY_DISCOVERY_PHASE=0 -I/Users/pschatzmann/Library/Arduino15/packages/arduino/hardware/mbed_nano/2.7.2/cores/arduino -I/Users/pschatzmann/Library/Arduino15/packages/arduino/hardware/mbed_nano/2.7.2/variants/ARDUINO_NANO33BLE -I/Users/pschatzmann/Library/Arduino15/packages/arduino/hardware/mbed_nano/2.7.2/cores/arduino/api/deprecated -I/Users/pschatzmann/Library/Arduino15/packages/arduino/hardware/mbed_nano/2.7.2/cores/arduino/api/deprecated-avr-comp -iprefix/Users/pschatzmann/Library/Arduino15/packages/arduino/hardware/mbed_nano/2.7.2/cores/arduino @/Users/pschatzmann/Library/Arduino15/packages/arduino/hardware/mbed_nano/2.7.2/variants/ARDUINO_NANO33BLE/includes.txt /Users/pschatzmann/Library/Arduino15/packages/arduino/hardware/mbed_nano/2.7.2/variants/ARDUINO_NANO33BLE/variant.cpp -o /var/folders/8q/1yw4v2l14s90kf739cqbc1rh0000gn/T/arduino_build_868723/core/variant.cpp.o
 

The I2S implementation in I2SNanoSenseBLE.h is currently checking for the define ARDUINO_ARDUINO_NANO33BLE

**Reply by BlueObsecurit** [reply]
I use the supplied core for the Xiao BLE sense, but I try (and in the case of the PWM example, successfully) add a define for it.
If I try to compile for nano 33 BLE core I get:

```
\src/AudioI2S/I2SStream.h: In member function 'virtual bool audio_tools::I2SStream::begin()':
\src/AudioI2S/I2SStream.h:87:30: error: void value not ignored as it ought to be
             return i2s.begin();
                              ^
\src/AudioI2S/I2SStream.h: In member function 'virtual int audio_tools::I2SStream::available()':
\src/AudioI2S/I2SStream.h:150:24: error: 'class audio_tools::I2SBase' has no member named 'available'
             return i2s.available();
                        ^~~~~~~~~
\src/AudioI2S/I2SStream.h: In member function 'virtual int audio_tools::I2SStream::availableForWrite()':
\src/AudioI2S/I2SStream.h:155:24: error: 'class audio_tools::I2SBase' has no member named 'availableForWrite'
             return i2s.availableForWrite();
                        ^~~~~~~~~~~~~~~~~
```
I assume I've broken something, as you can compile the player-sd-i2c example for Arduino BLE sense.
Tomorrow I will reinstall your library. I tried to get the Xiao BLE sense to be defined as Arduino BLE sense, but this didn't quite work. Perhaps I am using the wrong definition pulled from boards.txt? Is there a good way to easily have the library identify the Xiao board as the Arduino board?

**Comment by pschatzmann** [comment]
I committed some changes to support ARDUINO_SEEED_XIAO_NRF52840_SENSE

**Reply by BlueObsecurit** [reply]
It compiles! 
But I can't get the I2S dac to work.
I can compile SD to PWM player and hear something.
I've double checked everything, but it can still be something silly I've forgotten.
The big problem is that I can't get Serial to open on this sketch. Works with the SD to PWM but not this. So I can't tell what is happening.

I am using the Adafruit UDA1334A, and I've tried both normal and LSB mode.

**Reply by BlueObsecurit** [reply]
Thank you so much for adding support to your library.
I've had some more time to try to troubleshoot, I don't get very far.
I've tried various examples, all I managed to get out of the I2S DAC is a faint humming sound.
It seems to me that the microcontroller enters some kind of locked state. It is difficult to get it to reset, and difficult to open the serial console. All I get from the serial console is this about 10 ish times:

`[I] I2SNanoSenseBLE.h : 198 - frequency requested 7056000.000000 vs 125000.000000`

I've tried adding visual debugging (onboard LED, probably too crude), with no success.

**Comment by BlueObsecurit** [comment]
@pschatzmann 
I have spent some time trying to figure out what is happening, with limited success
Using the Streams Generator i2s example:
It seems the microcontroller stops working at some point during the i2s routine (?)
I have tried to add serial prints to give me an idea of when. I know that the Nano33BLE specific portion runs entirely.
But almost all attempts to add serial prints to the i2s stream results in not being able to open the serial console.
I fully accept that I am probably ill-equipped to know what to do, and that it would be better if I had some more knowledge/experience in order to know what is going on. Nevertheless, is this something you'd prefer not to deal with? If so let me know and I'll stop clogging up the discussion :)

**Comment by pschatzmann** [comment]
It seems that the microcontroller is crashing. 
W/o debugger you could add log statements with some long delay() to find the offending command.

**Comment by pschatzmann** [comment]
Just to give an update: the output is working now. I did not have any time to test the i2s input though
see https://github.com/pschatzmann/arduino-audio-tools/discussions/984
